runtime/dart_snapshot.cc

```diff
@@ -21,6 +21,22 @@ const char* DartSnapshot::kIsolateDataSymbol = "kDartIsolateSnapshotData";
 const char* DartSnapshot::kIsolateInstructionsSymbol =
     "kDartIsolateSnapshotInstructions";
 
+#if defined(OS_ANDROID)
```

Comment:
```
#if defined(OS_ANDROID)
#define SYMBOL_PREFIX "_"
#else
#defined SYMBOL_PREFIX ""
#endif
```

Followed by:

```
const char* DartSnapshot::kIsolateDataSymbol = SYMBOL_PREFIX "kDartIsolateSnapshotData";
```

This reads a lot clearer as the symbol name is only specified once in the string. You also don't need the extra constants for the SO variant.

Reply:
Also, on Android, when looking inside the currently loaded executable, we are still looking for kVMDataSymbol. So depending on whether the symbols are in the engine library (in debug JIT modes) vs the applications AOT library (in profile and release modes), there is an underscore? In that case, we should fix the inconsistency and make all symbol names be consistent.

Reply:
> Use SYMBOL_PREFIX

Good idea. Done.

> ... currently loaded executable, we are still looking for kVMDataSymbol

This is correct. There is a difference between the two. Because when the symbol is coming from C/C++ code we don't need the _ prefix, only if it comes from our generated `.S` file.

See here:

    % arm-linux-gnueabihf-nm apk/assets/app.so | grep kDartVmSnapshotData
    006a0e60 R _kDartVmSnapshotData

    % nm out/ReleaseX64/dart | grep kDartVmSnapshotData
    00000000002ba500 R kDartVmSnapshotData

Reply:
> In that case, we should fix the inconsistency and make all symbol names be consistent.

This is not necessarily in our control. As far as I could find out it works like this:

On Mac, the C++ compiler will prefix any C symbol `S` in the sources with a leading `_` when writing the symbol to the shared library. When doing a symbol lookup via `dlsym("S")` it will automatically prepend a leading underscore and look up `_S` instead.

On Linux, the C++ compiler will use the symbol as it is in C source code. So the shared library will have symbols without leading underscores.

So in order to make the Mac `dlsym()` succeed, the Dart VM writes out the symbols with leading underscores (see [runtime/vm/image_snapshot.cc](https://github.com/dart-lang/sdk/blob/0ad2b1243/runtime/vm/image_snapshot.cc#L308) ). But this requires us to prefix it on Linux when doing a `dlsym()`.

When the snapshots are embedded inside the executable, then there is [logic](https://github.com/dart-lang/sdk/blob/0ad2b1243d0fa9bfb1d8da270a593022ce050674/runtime/tools/bin_to_assembly.py#L57) to ensure to do the right prefixing. This is possible because we know exactly for which os/architecture the executable is.

Changing `dart_bootstrap` (which creates the `.S` file) to do prefixing based on target os/architecture would be a breaking change and might cause all kinds of issues.

I would really prefer if we could avoid it.

/cc @mraleph